cloudstate/eventsourced.go

```diff
@@ -0,0 +1,563 @@
+//
+// Copyright 2019 Lightbend Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cloudstate
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/cloudstateio/go-support/cloudstate/encoding"
+	"github.com/cloudstateio/go-support/cloudstate/protocol"
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/protobuf/ptypes/any"
+	"io"
+	"net/url"
+	"reflect"
+	"strings"
+	"sync"
+)
+
+// Entity
+type Entity interface {
+	EntityInitializer
+}
+
+// An EntityInitializer knows how to initialize an Entity
+type EntityInitializer interface {
+	New() interface{}
+}
```

Comment:
@marcellanz did you have any thoughts on this? I think it's one of the main warts of the API right now.

Reply:
I did. There are two points to it. One is that cloudstate somehow has to know how to create an entity out of a registered one. The other is that the Entity probably best knows how to create itself.

I aimed to have that not too cluttered or separated too much. The usage of the "nil pointer" seems to used by other libraries and here the gRPC lib itself (see the register functions). In Java its easier as we have a class to be used, but not eith go.

Either way, I wanted to have a most type safe and simple way to register an entity and also one where the user is guided like by implementing an interface and not providing a string or the like. What other ways of doing such a registration following such principles do you see? Having a factory with EventSourcedEntity as you stated would fit into that pattern I think.

Reply:
In OO languages you would usually have a static method to create instances (rather than an instance create itself). In golang we just have a function (without a receiver) to create instances. That is the most natural I think. Right now the requirement to use an instance of the Entity to create itself feels like an unnecessary complication - it leads to things like:

```
func (sc ShoppingCart) New() interface{} {
	return NewShoppingCart()
}
```

Where the receiver `sc` isn't used. Instead I think we should just pass `NewShoppingCart` directly:

```
	err = cloudState.RegisterEventSourcedEntity(
		&cloudstate.EventSourcedEntity{
			EntityFn:      NewShoppingCart,
			ServiceName: "com.example.shoppingcart.ShoppingCart",
		},

```

Functions are first class citizens and this feels like a more natural way to me to pass a factory method.

Reply:
Looks good, let me try that.

Reply:
So I've implemented this and it is way cleaner and much less code, obviously. At the same time it enables compile time enforcement to implement command and event handlers. Great!